### PR TITLE
TLS certificate is added to the ingress for the eventlistener

### PR DESCRIPTION
**Pull request by akihikokuroda**
This PR for https://github.com/tektoncd/experimental/issues/279

when the callback URL has "https" protocol.
The name of the TLS certificate secret can be specified in WEBHOOK_TLS_CERTIFICATE
environment variable.  The default is "cert-" + eventlistener name.
When the secret doesn't exit with the name, the new certificate secret
is created and the certificate is signed by the framework.

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by akihikokuroda** [comment]
@dibbles Thanks for review.  I added statements in the install docs.

File: webhooks-extension/pkg/endpoints/webhook.go [hunk 1]
```diff
@@ -462,7 +471,7 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
 				RespondError(response, errors.New(msg), http.StatusInternalServerError)
 				return
 			} else {
-				logging.Log.Debug("ingress creation taskrun succeeded")
+				logging.Log.Debug("ingress creation succeeded")
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/webhook.go]: The route automatically gets the certificate signed by the platform.

File: webhooks-extension/pkg/endpoints/webhook.go [hunk 2]
```diff
@@ -580,6 +613,90 @@ func (r Resource) createDeleteIngress(mode, installNS string) error {
...
+func (r Resource) createCertificate(secretName, installNS, callback string) string {
+	var key, crt []byte
+
+	priv, _ := rsa.GenerateKey(rand.Reader, 2048)
+	template := x509.CertificateRequest{
+		Subject: pkix.Name{
+			CommonName:   callback,
+			Country:      []string{"US"},
+			Province:     []string{"NC"},
+			Organization: []string{"IBM"},
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/webhook.go]: I change these with generic strings.  This certificate should be replaced by the user if it is used in the production.  The user certificate secret should be set in the  WEBHOOK_TLS_CERTIFICATE to do so.

File: webhooks-extension/pkg/endpoints/webhook.go [hunk 3]
```diff
@@ -580,6 +613,90 @@ func (r Resource) createDeleteIngress(mode, installNS string) error {
...
+	// approve csr manually
+	csrRecord, err = client.Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		logging.Log.Errorf("Failed getting CSR record: %v", err)
+		return ""
+	}
+	csrRecord.Status.Conditions = append(csrRecord.Status.Conditions, certv1beta1.CertificateSigningRequestCondition{
+		Type:    certv1beta1.CertificateApproved,
+		Reason:  "AutoApproved",
+		Message: "Approved by Tekton webhook",
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/webhook.go]: I don't see issue here.

File: webhooks-extension/pkg/endpoints/webhook.go [hunk 4]
```diff
@@ -580,6 +613,90 @@ func (r Resource) createDeleteIngress(mode, installNS string) error {
...
+	})
+	_, err = client.UpdateApproval(csrRecord)
+	if err != nil {
+		logging.Log.Errorf("Failed approving CSR: %v", err)
+		return ""
+	}
+
+	csrdata, err = csr.WaitForCertificate(client, csrRecord, 50000000)
+	if err != nil {
+		logging.Log.Errorf("Failed waiting certificate: %v", err)
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/webhook.go]: I fix it.

File: webhooks-extension/pkg/endpoints/webhook.go [hunk 5]
```diff
@@ -580,6 +613,90 @@ func (r Resource) createDeleteIngress(mode, installNS string) error {
...
+		Reason:  "AutoApproved",
+		Message: "Approved by Tekton webhook",
+	})
+	_, err = client.UpdateApproval(csrRecord)
+	if err != nil {
+		logging.Log.Errorf("Failed approving CSR: %v", err)
+		return ""
+	}
+
+	csrdata, err = csr.WaitForCertificate(client, csrRecord, 50000000)
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/webhook.go]: it's nano second.  Actually it's too short so I increase it.

File: webhooks-extension/pkg/endpoints/webhook.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/webhook.go]: The route automatically gets the certificate signed by the platform.

**Comment by akihikokuroda** [comment]
Should I resolve the merge conflict now or later?

**Comment by akihikokuroda** [comment]
Retest after rebase seems OK.

**Comment by mnuttall** [comment]
Hi @akihikokuroda - sorry to have to ask - one more rebase please?

**Comment by akihikokuroda** [comment]
@mnuttall rebase is done.

**Pull request merged**